### DISCUSSION TITLE
Test against over spending

### DISCUSSION THREAD

**Discussion by pierrekttipay** [codenotary/immudb#2024] [Q&A] [answered]
As a proof of concept. I would like to write some client code to test against double spending. Let me explain.

I would like to have this sequence of 3 events happen at a massive scale with the last 2 competing with each other:
1) OpenAccountX balance 100
2) Withdraw 50 from AccountX
3) Withdraw 60 from AccountX
each time creating new Account (incrementing X) and repeating this over and over again.

My worry is if event-2 lookup the balance, see 100 while event-3 happen to do the same thing at the same time and see 100 as well. I want to test that one of the transaction will actually fail and will not end up with a negative balance.

In order to prevent that, the client will have to check if the balance will be > 0 after executing the withdraw and reject the transaction if otherwise.
Because the event-2 and event-3 compete at the same time. Some account will have a balance of 50, some will have a balance of 40 (because event-3 was applied before event-2 in that occurance)

Can someone please help me model this simple problem in either: Go, Python or Node please?

**Comment by ostafen** [comment]
Hey, @pierrekttipay, thanks for asking and opening the discussion.
Would you like to model the problem using the SQL or the key-value layer?

**Reply by pierrekttipay** [reply]
Hey @ostafen 
Thank you for your quick response.
I would prefer SQL for this example.
Thanks for helping

**Reply by ostafen** [reply, marked as answer]
The following snippets creates an `accounts` table with two simple fields: id and balance. Then one account is inserted, with an initial balance of `65`. Note the `CHECK balance >= 0` part, which makes sure no update can result in a negative balance.

The snippet then spawns `10` concurrent goroutines, each trying to decrement the balance by `10` (they try as soon as a read conflict happens). If a unique constraint violation is detected the goroutine simply stops, as it would otherwise loop forever.

If all works, we expect the balance to be decremented at most `6` times, with a final balance of value of `5`.

```golang
package main

import (
	"context"
	"errors"
	"fmt"
	"log"
	"sync"

	"github.com/codenotary/immudb/embedded/sql"
	"github.com/codenotary/immudb/embedded/store"
)

func fatalOnErr(err error) {
	if err != nil {
		log.Fatal(err)
	}
}

func main() {
	st, err := store.Open("./data", store.DefaultOptions().WithMultiIndexing(true))
	fatalOnErr(err)
	defer st.Close()

	e, err := sql.NewEngine(st, sql.DefaultOptions())
	fatalOnErr(err)

	ctx := context.Background()
	_, _, err = e.Exec(ctx, nil, `
		CREATE TABLE IF NOT EXISTS accounts(
			id INTEGER, 
			balance INTEGER, 
			
			CHECK (balance >= 0),

			PRIMARY KEY id
		)
	`, nil)
	fatalOnErr(err)

	_, _, err = e.Exec(ctx, nil, "UPSERT INTO accounts(id, balance) VALUES (1, 65)", nil)
	fatalOnErr(err)

	n := 10
	var wg sync.WaitGroup
	wg.Add(n)

	for i := 0; i < n; i++ {
		go func() {
			defer wg.Done()

			err := store.ErrTxReadConflict
			for err == store.ErrTxReadConflict {
				_, _, err = e.Exec(ctx, nil, "UPDATE accounts SET balance = balance - 10", nil)
			}

			if err != nil && !errors.Is(err, sql.ErrCheckConstraintViolation) {
				fatalOnErr(err)
			}
		}()
	}
	wg.Wait()

	reader, err := e.Query(ctx, nil, "SELECT * FROM accounts", nil)
	fatalOnErr(err)

	row, err := reader.Read(ctx)
	fatalOnErr(err)

	balance, _ := row.ValuesByPosition[1].RawValue().(int64)
	if balance != 5 {
		fatalOnErr(fmt.Errorf("unexpected balance: %d", balance))
	}
	fatalOnErr(reader.Close())
}
```

@pierrekttipay: let me know if that is helpful and if you need further assistance. Thanks!

**Reply by pierrekttipay** [reply]
Thanks @ostafen , I will check the response in detail.
The technique you used is fascinating. You used `CHECK (balance >= 0)` 
I couldn't read anything about this CHECK functionality in the immudb documentation.
1) Is there something similar for KV? 
2) Is there some other way to protect but at the statement UPDATE level instead of the TABLE CHECK level? My thinking is if some accounts are allowed $100 credit, $200 credits and other can not go into credit at all.
I guess what I am asking is: is there a way to use TRANSACTION to protect ourselves at the UPDATE level? How would it looks like, please?

I think this would work well with interfacing with real 3rd party payment system. For instance:
- the payment will ask us if the customer has enough balance. We run the UPDATE and it worked so we authorise the payment.
- if for whatever reason the payment fail, we will be notified and re-credit the account.
- if other payments are happening at the same time on the same account the CHECK will make sure we reject the transactions that over spend.

**Comment by pierrekttipay** [comment]
Hi @ostafen 
Thank you again for your example. 
I have been modifying it to call an external DB instance instance of using an embedded one.
The embedded one is not strong enough for my POC.
I have been able to progress, but I am facing a massive amount of `tx read conflict`. I was hoping you could help with making my code retry failed tx read hoping the retry will succeed.

my code is:
```go
package main

import (
	"context"
	"fmt"
	"log"
	"sync"

	immudb "github.com/codenotary/immudb/pkg/client"
)

func fatalOnErr(err error) {
	if err != nil {
		log.Fatal(err)
	}
}

func main() {

	opts := immudb.DefaultOptions().WithAddress("127.0.0.1").WithPort(3322)

	e := immudb.NewClient().WithOptions(opts)

	// connect with immudb server (user, password, database)
	err := e.OpenSession(context.Background(), []byte("immudb"), []byte("immudb"), "defaultdb")
	if err != nil {
		log.Fatal(err)
	}
	// ensure connection is closed
	defer e.CloseSession(context.Background())

	ctx := context.Background()

	if err != nil {
		log.Fatalf("Load Table %s. Error while creating transaction", err)
	}

	_, err = e.SQLExec(ctx, `
		CREATE TABLE IF NOT EXISTS accounts(
			id INTEGER, 
			balance INTEGER, 
			
			CHECK (balance >= 0),

			PRIMARY KEY id
		)
	`, nil)
	fatalOnErr(err)

	_, err = e.SQLExec(ctx, "UPSERT INTO accounts(id, balance) VALUES (1, 65)", nil)
	fatalOnErr(err)
	log.Printf("UPSERT $65")

	n := 10
	var wg sync.WaitGroup
	wg.Add(n)

	for i := 0; i < n; i++ {
		go func() {
			defer wg.Done()

			tx, err := e.NewTx(ctx)
			if err != nil {
				fatalOnErr(err)
			}
			err = tx.SQLExec(ctx, "UPDATE accounts SET balance = balance - 10", nil)
			if err != nil {
				fatalOnErr(err)
			}
			log.Printf("UPDATE balance - $10")
			_, err = tx.Commit(ctx)
			if err != nil {
				log.Printf("%v", err)
				// fatalOnErr(err)
			}
		}()
	}
	wg.Wait()

	res, err := e.SQLQuery(ctx, "SELECT * FROM accounts", map[string]interface{}{}, true)
	fatalOnErr(err)

	for _, row := range res.Rows {
		log.Printf("Got row: %v\n", row)
		vals := row.GetValues()
		log.Printf("Got values: %v\n", vals)
		balance := vals[1].GetN()
		if balance != 5 {
			fatalOnErr(fmt.Errorf("unexpected balance: %d", balance))
		}
	}
}
```
my output varies. Sometimes balances is $45, sometimes $55. (starting with a balance of $65)
Basically only the first `UPDATE` was able to succeed.
Here is my output:
```
(base) ➜  balance-check go run main.go
2024/10/12 17:39:23 UPSERT $65
2024/10/12 17:39:23 UPDATE balance - $10
2024/10/12 17:39:23 UPDATE balance - $10
2024/10/12 17:39:23 UPDATE balance - $10
2024/10/12 17:39:23 UPDATE balance - $10
2024/10/12 17:39:23 UPDATE balance - $10
2024/10/12 17:39:23 UPDATE balance - $10
2024/10/12 17:39:23 UPDATE balance - $10
2024/10/12 17:39:23 UPDATE balance - $10
2024/10/12 17:39:23 UPDATE balance - $10
2024/10/12 17:39:23 UPDATE balance - $10
2024/10/12 17:39:23 tx read conflict
2024/10/12 17:39:23 tx read conflict
2024/10/12 17:39:23 tx read conflict
2024/10/12 17:39:23 tx read conflict
2024/10/12 17:39:23 tx read conflict
2024/10/12 17:39:23 tx read conflict
2024/10/12 17:39:23 tx read conflict
2024/10/12 17:39:23 tx read conflict
2024/10/12 17:39:23 tx read conflict
2024/10/12 17:39:23 Got row: columns:"(accounts.id)"  columns:"(accounts.balance)"  values:{n:1}  values:{n:55}
2024/10/12 17:39:23 Got values: [n:1 n:55]
2024/10/12 17:39:23 unexpected balance: 55
exit status 1
```
If I try to extrapolate to a real scenario. I will have many customers trying to withdraw money at the same time.
I want to allow them as long as there is enough balance. 
Here I have enough but I simply fail processing their request because of a `tx read conflict` 
Can you please help finding a way to retry update after a `tx read conflict` ?

**Reply by ostafen** [reply]
Hey, @pierrekttipay, I think you can just retry the update in a loop (bounded by some maximum number of retries) and add a small random delay (or exponential backoff delay) when you get a read conflict.

**Reply by pierrekttipay** [reply]
I am struggling a little with my loop
```go
	for i := 0; i < n; i++ {
		go func() {
			defer wg.Done()
			_, err := updateAccount(ctx, e)

			log.Printf("+%s", err)
			for err != nil {
				time.Sleep(1 * time.Nanosecond)
				_, err = updateAccount(ctx, e)
				log.Printf("-%v", err)
				// fatalOnErr(err)
			}

		}()
	}
```
my conditional `err != nil` is too simplistic and will create an infinite loop when the balance will reach a negative number.
`check constraint violation: (balance >= 0)` 
I need to be able to test the error type and stop the loop if the err is about constraint check

My updateAccount function is as such:
```go
func updateAccount(ctx context.Context, e immudb.ImmuClient) (string, error) {
	tx, err := e.NewTx(ctx)
	if err != nil {
		fatalOnErr(err)
	}
	err = tx.SQLExec(ctx, "UPDATE accounts SET balance = balance - 10", nil)
	if err != nil {
		return "", err
	}
	log.Printf("UPDATE balance - $10")
	_, err = tx.Commit(ctx)
	if err != nil {
		log.Printf("-%v", err)
		// fatalOnErr(err)
	}
	return "", err
}
```
I could only find the `ErrCheckConstraintViolation` in the `embedded` sql package which tells me it's probably not the one I need here.
How would you write this test to break the loop if the error is about constraint violation, please?

**Reply by pierrekttipay** [reply]
For instance, doing this 
```go
			log.Printf("+%s", err)
			for err != nil && !errors.Is(err, sql.ErrCheckConstraintViolation) {
				time.Sleep(1 * time.Nanosecond)
				_, err = updateAccount(ctx, e)
				log.Printf("-%v", err)
				// fatalOnErr(err)
			}
```
will not avoid the infinite loop
```
2024/10/12 21:32:50 -check constraint violation: (balance >= 0)
2024/10/12 21:32:50 -check constraint violation: (balance >= 0)
2024/10/12 21:32:50 -check constraint violation: (balance >= 0)
2024/10/12 21:32:50 -check constraint violation: (balance >= 0)
2024/10/12 21:32:50 -check constraint violation: (balance >= 0)
2024/10/12 21:32:50 -check constraint violation: (balance >= 0)
2024/10/12 21:32:50 -check constraint violation: (balance >= 0)
2024/10/12 21:32:50 -check constraint violation: (balance >= 0)
2024/10/12 21:32:50 -check constraint violation: (balance >= 0)
2024/10/12 21:32:50 -check constraint violation: (balance >= 0)
...
```
here is a `log.Printf("%w", err)`

```
%!w(*errors.immuError=&{ XX000 check constraint violation: (balance >= 0) 0 })
```

**Reply by ostafen** [reply]
Because the error returned from the remote server is wrapped in an immuError error.

try:

```golang
strings.Contains(err.Error(), sql.ErrCheckConstraintViolation.Error())
```

**Reply by pierrekttipay** [reply]
Ok this worked. Thanks
```
			for err != nil && !strings.Contains(err.Error(), sql.ErrCheckConstraintViolation.Error()) {
				time.Sleep(1 * time.Nanosecond)
				_, err = updateAccount(ctx, e)
				// fatalOnErr(err)
			}
```
I would have thought there will be a more idiomatic way of doing it using `errors.As` and its unwrap capability.

**Comment by pierrekttipay** [comment]
Hi again @ostafen 
I have been playing with the code and tried to remove the constraint and the table level and recreate it at the transaction level.
This way I was able to return a custom `ErrNotEnoughBalance`.
 
Do you see any downside with that?
```go
package main

import (
	"context"
	"errors"
	"fmt"
	immudb "github.com/codenotary/immudb/pkg/client"
	"log"
	"sync"
	"time"
)

var (
	id                  int = 3
	ErrNotEnoughBalance     = errors.New("not enough balance")
)

func fatalOnErr(err error) {
	if err != nil {
		log.Fatal(err)
	}
}

func updateAccount(ctx context.Context, e immudb.ImmuClient) error {

	tx, err := e.NewTx(ctx)
	if err != nil {
		return err
	}
	res, err := tx.SQLQuery(ctx, fmt.Sprintf("SELECT balance FROM accounts2 WHERE id=%v", id), map[string]interface{}{})
	fatalOnErr(err)

	if res.Rows[0].GetValues()[0].GetN() > 10 {
		err = tx.SQLExec(ctx, fmt.Sprintf("UPDATE accounts2 SET balance = balance - 10 WHERE id=%v", id), nil)
		if err != nil {
			return err
		}
		log.Printf("UPDATE balance - $10")
		_, err = tx.Commit(ctx)
		if err != nil {
			log.Printf("%v", err)
		}
	} else {
		err = tx.Rollback(ctx)
		if err != nil {
			log.Printf("%v", err)
		}
		return ErrNotEnoughBalance
	}
	return err
}

func main() {

	opts := immudb.DefaultOptions().WithAddress("127.0.0.1").WithPort(3322)

	e := immudb.NewClient().WithOptions(opts)

	// connect with immudb server (user, password, database)
	err := e.OpenSession(context.Background(), []byte("immudb"), []byte("immudb"), "defaultdb")
	if err != nil {
		log.Fatal(err)
	}
	// ensure connection is closed
	defer e.CloseSession(context.Background())

	ctx := context.Background()

	if err != nil {
		log.Fatalf("Load Table %s. Error while creating transaction", err)
	}

	_, err = e.SQLExec(ctx, `
		CREATE TABLE IF NOT EXISTS accounts2(
			id INTEGER, 
			balance INTEGER, 
			
			PRIMARY KEY id
    )	
	`, nil)
	fatalOnErr(err)

	_, err = e.SQLExec(ctx, fmt.Sprintf("UPSERT INTO accounts2(id, balance) VALUES (%v, 65)", id), nil)
	fatalOnErr(err)
	log.Printf("UPSERT $65")

	n := 10
	var wg sync.WaitGroup
	wg.Add(n)

	for i := 0; i < n; i++ {
		go func() {
			defer wg.Done()
			success := false
			err := updateAccount(ctx, e)

			for err != nil && !errors.Is(err, ErrNotEnoughBalance) {
				time.Sleep(1 * time.Nanosecond)
				err = updateAccount(ctx, e)
			}
			if err == nil {
				success = true
			}
			log.Printf("STOP success:%v", success)

		}()
	}
	wg.Wait()

	log.Printf("ALL FINISHED")
	res, err := e.SQLQuery(ctx, fmt.Sprintf("SELECT * FROM accounts2 WHERE id=%v", id), map[string]interface{}{}, true)
	fatalOnErr(err)

	for _, row := range res.Rows {
		log.Printf("Got row: %v\n", row)
		vals := row.GetValues()
		log.Printf("Got values: %v\n", vals)
		balance := vals[1].GetN()
		if balance != 5 {
			fatalOnErr(fmt.Errorf("unexpected balance: %d", balance))
		}
	}
}
```

**Comment by ostafen** [comment]
It's totally fine, but I would use the check syntax unless you need to implement more complex checks

**Comment by pierrekttipay** [comment]
I marked as resolved, thank you very much.
If you think about it, please ping me here again if error handling is improved in the future (as comparing strings doesn't seem right)
Thanks
